### PR TITLE
Feature/blind vesting

### PR DESCRIPTION
**Pull request by malteish**
@Christoph says:

> This allows for priavte vesting. Only the total combined amount of tokens to be vested is public through the minting allowance.
> 
> The company hashes the vesting terms and a 32 bytes secret together and commits it to the vesting contract. The beneficiary can then claim his vesting by calling createAward, revealing the terms and the secret.

### CODE REVIEW

File: src/DssVest.sol [hunk 1]
```diff
@@ -174,6 +176,48 @@ abstract contract DssVest is ERC2771Context {
...
+        @dev Governance adds a vesting contract
+        @param _usr The recipient of the reward
+        @param _tot The total amount of the vest
+        @param _bgn The starting timestamp of the vest
+        @param _tau The duration of the vest (in seconds)
+        @param _eta The cliff duration in seconds (i.e. 1 years)
+        @param _mgr An optional manager for the contract. Can yank if vesting ends prematurely.
+        @return id  The id of the vesting contract
+    */
+    function create(address _usr, uint256 _tot, uint256 _bgn, uint256 _tau, uint256 _eta, address _mgr) external returns (uint256 id) {
```
Comment (CJentzsch) [inline src/DssVest.sol]: missing `lock` modifier

File: src/DssVest.sol [hunk 2]
```diff
@@ -174,6 +176,48 @@ abstract contract DssVest is ERC2771Context {
...
+        @param _eta The cliff duration in seconds (i.e. 1 years)
+        @param _mgr An optional manager for the contract. Can yank if vesting ends prematurely.
+        @param _slt The salt used to increase privacy when committing
+        @return id  The id of the vesting contract
+    */
+    function createFromCommitment(bytes32 _bch, address _usr, uint256 _tot, uint256 _bgn, uint256 _tau, uint256 _eta, address _mgr, bytes32 _slt) external lock returns (uint256 id) {
+        require(_bch == keccak256(abi.encodePacked(_usr, _tot, _bgn, _tau, _eta, _mgr, _slt)), "DssVest/invalid-hash");
+        require(commitments[_bch], "DssVest/commitment-not-found");
+        id = _create(_usr, _tot, _bgn, _tau, _eta, _mgr);
+        commitments[_bch] = false;
```
Comment (CJentzsch) [inline src/DssVest.sol:217]: it is safer to set `commitments[_bch] = false;` before `create` is executed
, right after it was checked that the commit exists.

Comment (CJentzsch) [reply src/DssVest.sol:217]: (this is for protecting against reentrancy, but I know `lock` is already doing that. It is just a second layer of security, technically not needed.

File: src/DssVest.sol [hunk 3]
```diff
@@ -174,6 +176,48 @@ abstract contract DssVest is ERC2771Context {
         require((z = uint128(x)) == x, "DssVest/uint128-overflow");
     }
 
+    /** 
+        @dev commit to the creation of an award without revealing the award's contents yet
+        @param bch  Blind Commitment Hash - The hash of the award's contents, see hash function in createAward for details
+    */
+    function commit(bytes32 bch) external lock auth {
+        commitments[bch] = true;
```
Comment (CJentzsch) [inline src/DssVest.sol:198]: we should emit an event here

**Review by CJentzsch — CHANGES_REQUESTED** [review]

File: src/DssVest.sol (hunk 1 above)
Comment (malteish) [reply src/DssVest.sol]: good catch [e9571cd](https://github.com/corpus-io/dss-vest/pull/25/commits/e9571cdd990cf7110bfcf9d97ad1c3ae00d9f03a)

File: src/DssVest.sol (hunk 2 above)
Comment (malteish) [reply src/DssVest.sol:217]: done [b282ab2](https://github.com/corpus-io/dss-vest/pull/25/commits/b282ab2866f07ef33632d00eda49747be1bf5842)

File: src/DssVest.sol (hunk 3 above)
Comment (malteish) [reply src/DssVest.sol:198]: true emitting the claim, too. [082a2bc](https://github.com/corpus-io/dss-vest/pull/25/commits/082a2bc5236f13c7c821a099fa203cb7045d5e41)

**Review by CJentzsch — APPROVED** [review]

**Pull request merged**